### PR TITLE
[SPARK-7133][SQL] Implement struct, array, and map field accessor

### PR DESCRIPTION
**Pull request by cloud-fan**
It's the first step: generalize UnresolvedGetField to support all map, struct, and array
TODO: add `apply` in Scala and `__getitem__` in Python, and unify the `getItem` and `getField` methods to one single API(or should we keep them for compatibility?).

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @marmbrus , it should be easy to add apply method, so I'll do it after code review.

**Comment by liancheng** [comment]
ok to test

**Comment by rxin** [comment]
We should definitely keep the old function for compatibility.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 1]
```diff
@@ -194,7 +194,7 @@ case class ResolvedStar(expressions: Seq[NamedExpression]) extends Star {
   override def toString: String = expressions.mkString("ResolvedStar(", ", ", ")")
 }
 
-case class UnresolvedGetField(child: Expression, fieldName: String) extends UnaryExpression {
+case class UnresolvedGetField(child: Expression, fieldExpr: Expression) extends UnaryExpression {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: can you add some javadoc for this class. at the very least, we should say this can be used to get a field out of struct/map/array.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  override def eval(input: Row): Any = {
+    val baseValue = child.eval(input).asInstanceOf[Row]
+    if (baseValue == null) null else baseValue(ordinal)
+  }
+}
+
+/**
+ * Returns the array of value of fields in the Array of Struct `child`.
+ */
+case class ArrayStructGetField(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: i'm sort of confused - why is this needed?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: see https://issues.apache.org/jira/browse/SPARK-2096.
"arrayOfStruct.field1" means to access all values of "field1" in this array of structs and the returns those values as an array.

**Comment by cloud-fan** [comment]
The failed test cases looks unrelated, retest?

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, add to whitelist.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 3]
```diff
@@ -203,5 +210,5 @@ case class UnresolvedGetField(child: Expression, fieldName: String) extends Unar
   override def eval(input: Row = null): EvaluatedType =
     throw new TreeNodeException(this, s"No function to evaluate expression. type: ${this.nodeName}")
 
-  override def toString: String = s"$child.$fieldName"
+  override def toString: String = s"$child.getField($fieldExpr)"
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: to make it shorter, maybe just "$child[$fieldExpr]".

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      case (ArrayType(StructType(fields), containsNull), Literal(fieldName, StringType)) =>
+        val ordinal = findField(fields, fieldName.toString, resolver)
+        ArrayStructGetField(child, fields(ordinal), ordinal, containsNull)
+      case (_: ArrayType, _) if fieldExpr.dataType.isInstanceOf[IntegralType]  =>
+        ArrayOrdinalGetField(child, fieldExpr)
+      case (_: MapType, _) =>
+        MapOrdinalGetField(child, fieldExpr)
+      case (otherType, _) =>
+        throw new AnalysisException(
+          "GetField is not valid on child of type " +
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: I think we'd want a more friendly error message for the end user of SQL / data frames. I don't think know much about GetField, which is an internal concept.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Previously we use `$child[$fieldExpr]` to represent `GetItem`(for Map and Array) and use `$child.$fieldName` to represent `GetField`(for Struct), now they are combined, I have no idea how to name it...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis._
+import org.apache.spark.sql.types._
+
+object GetField {
+  /**
+   * Returns the resolved `GetField`. It will return one kind of concrete `GetField`,
+   * depend on the type of `child` and `fieldExpr`.
+   */
+  def apply(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: we should document the logic here, i.e. in what case do we use which.

The code is still somewhat harder to read.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala (hunk 4 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: one way I can think of to make the error messages more friendly is to specialize it for each child.datatype.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+      case (_: MapType, _) =>
+        MapOrdinalGetField(child, fieldExpr)
+      case (otherType, _) =>
+        throw new AnalysisException(
+          "GetField is not valid on child of type " +
+            s"$otherType with fieldExpr of type ${fieldExpr.dataType}")
+    }
+  }
+
+  def unapply(g: GetField): Option[(Expression, Expression)] = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: what is the unapply semantics?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]:
```
$child[$fieldExpr]
```
should be fine.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    g match {
+      case _: StructGetField => Some((g.child, null))
+      case o: OrdinalGetField => Some((o.child, o.ordinal))
+      case _ => None
+    }
+  }
+
+  /**
+   * find the ordinal of StructField, report error if no desired field or over one
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: uppercase Find

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: The semantics is to return expressions the `GetField` have, it's used in [Optimizer](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L228-L231)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  }
+}
+
+trait GetField extends UnaryExpression {
+  self: Product =>
+
+  type EvaluatedType = Any
+}
+
+abstract class StructGetField extends GetField {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: maybe we can remove this? I don't see a lot of benefits of having a common parent type for these 2 classes. On the contrary it makes the type hierarchy fairly complicated.

**Comment by rxin** [comment]
The changes looks pretty good. Left comments mostly about documentation, naming, and error messages.

**Comment by cloud-fan** [comment]
Jenkins retest it please.

**Comment by cloud-fan** [comment]
Retest this please.

File: python/pyspark/sql/dataframe.py [hunk 9]
```diff
@@ -1188,13 +1188,13 @@ def getField(self, name):
         >>> from pyspark.sql import Row
         >>> df = sc.parallelize([Row(r=Row(a=1, b="b"))]).toDF()
         >>> df.select(df.r.getField("b")).show()
-        r.b
+        r[b]
         b
         >>> df.select(df.r.a).show()
-        r.a
+        r[a]
```
Comment (cloud-fan) [inline python/pyspark/sql/dataframe.py]: It really surprise me that these doc-like String will be run as test case...

Comment (rxin) [inline python/pyspark/sql/dataframe.py]: yup Python is magical.

**Comment by rxin** [comment]
I will let @marmbrus take a look at this tomorrow.

Meantime, can you add the apply method and Python getitem method? Thanks.

**Comment by cloud-fan** [comment]
@rxin , already done :)

File: python/pyspark/sql/dataframe.py [hunk 10]
```diff
@@ -1166,7 +1166,7 @@ def __init__(self, jc):
 
     # container operators
     __contains__ = _bin_op("contains")
-    __getitem__ = _bin_op("getItem")
+    __getitem__ = _bin_op("apply")
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: can we add a unit test?

you can add it in tests.py

**Comment by cloud-fan** [comment]
Retest this please.

**Comment by rxin** [comment]
@cloud-fan mind bringing this up to date? I will ask @marmbrus to review this soon.

**Comment by cloud-fan** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 11]
```diff
@@ -68,6 +68,16 @@ class Column(protected[sql] val expr: Expression) extends Logging {
   override def hashCode: Int = this.expr.hashCode
 
   /**
+   * An expression that gets an item at a position out of an [[ArrayType]],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Extracts a value or values from a complex type.  The following types of extraction are supported:
 - Given an `Array`, an integer ordinal can be used to retrieve a single value.
 - Given a `Struct`, a string fieldname can be used to extract that field.
 - Given an `Array` of `Struct`s, a string can be used to "map" the array to a new array containing only the specified field from each `Struct`.
 - Given a `Map`, a key of the correct type can be used to retrieve an individual value.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+package org.apache.spark.sql.catalyst.expressions
+
+import scala.collection.Map
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis._
+import org.apache.spark.sql.types._
+
+object GetField {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala]: I have some naming suggestions here:

 - `(Unresolved)GetField` => `(Unresolved)ExtractValue`
 - `SimpleStructGetField` => `GetStructField`
 - `ArrayStructGetField` => `MapStructArray`?
 - `ArrayOrdinalGetField` => `GetArrayItem`
 - `MapOrdinalGetField` => `GetMapValue`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 13]
```diff
@@ -43,7 +43,7 @@ abstract class DataType {
   /**
    * Enables matching against DataType for expressions:
    * {{{
-   *   case Cast(child @ BinaryType(), StringType) =>
+   *   case Cast(child @ DataType(), StringType) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: I'm not sure we actually want to make this change.  This is just an illustration of a case where this function would be used.  Matching on `DataType` itself is not possible, right?

**Comment by marmbrus** [comment]
Other than naming and documentation this LGTM.

**Comment by marmbrus** [comment]
Thanks!  Merged to master and 1.4.
